src/visidentifier/pocowsclient.hpp

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (C) 2024 Renesas Electronics Corporation.
+ * Copyright (C) 2024 EPAM Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef POCOWSCLIENT_HPP_
+#define POCOWSCLIENT_HPP_
+
+#include <memory>
+#include <mutex>
+
+#include <Poco/Event.h>
+#include <Poco/Net/HTTPMessage.h>
+#include <Poco/Net/HTTPRequest.h>
+#include <Poco/Net/HTTPResponse.h>
+#include <Poco/Net/HTTPSClientSession.h>
+#include <Poco/Net/WebSocket.h>
+
+#include "visidentifier/wsclient.hpp"
+#include "wsclientevent.hpp"
+#include "wspendingrequests.hpp"
+
+/**
+ * Poco web socket client.
+ */
+class PocoWSClient : public WSClientItf {
+public:
+    /**
+     * Creates Web socket client instance.
+     *
+     * @param config web socket config.
+     */
+    PocoWSClient(const WSConfig& config);
+
+    /**
+     * Connects to Web Socket server.
+     */
+    void Connect() override;
+
+    /**
+     * Closes Web Socket client.
+     */
+    void Close() override;
+
+    /**
+     * Disconnects Web Socket client.
+     */
+    void Disconnect() override;
+
+    /**
+     * Generates request id.
+     *
+     * @returns std::string
+     */
+    std::string GenerateRequestID() override;
+
+    /**
+     * Waits for Web Socket Client Event.
+     *
+     * @returns WSClientEvent::Details
+     */
+    WSClientEvent::Details WaitForEvent() override;
+
+    /**
+     * Sets message handler.
+     *
+     * @param handler handler functor
+     * @returns WSClientEvent&
+     */
+    void SetMessageHandler(MessageHandlerFunc handler) override;
+
+    /**
+     * Sends request. Blocks till the response is received or timed-out (WSException is thrown).
+     *
+     * @param requestId request id
+     * @param message request payload
+     * @returns ByteArray
+     */
+    ByteArray SendRequest(const std::string& requestId, const ByteArray& message) override;
+
+    /**
+     * Sends message. Doesn't wait for response.
+     *
+     * @param message request payload
+     */
+    void SendMessage(const ByteArray& message);
+
+    /**
+     * Destroys web socket client instance.
+     */
+    ~PocoWSClient() override;
+
+private:
+    void HandleResponse(const std::string& frame);
+    void ReceiveFrames();
+    void StartReceiveFramesThread();
+    void StopReceiveFramesThread();
+
+    WSConfig                                       mWSConfig;
+    std::recursive_mutex                           mMutex;
+    std::unique_ptr<std::thread>                   mReceivedFramesThread;
+    bool                                           mIsConnected {false};
+    Poco::SharedPtr<Poco::Net::Context>            mCryptoContext;
+    std::unique_ptr<Poco::Net::HTTPSClientSession> mClientSession;
```

Comment:
why we need to use std::unique_ptr

Reply:
HTTPSClientSession is not copyable or movable
unique_ptr is the best option on my opinion.

Reply:
Why we need allocate memory? Who will be copy this object ? if you mean copy  PocoWSClient than make copy constructor as delete. For my opinion this not good without a reason allocate dyn memory

Reply:
Poco::Net::HTTPSClientSession interface allows connection to service only in the constructor

Reply:
Failed to use optional